FR #962: Add Correction Handling and Delivery Handling attributes and refactor Invoice Handling (#1501)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/columns/chargeclass.md

# Charge Class

Charge Class indicates whether a [*charge*](#glossary:charge) represents a [*correction](#glossary:correction) to a previously [*closed billing period*](#glossary:closed-billing-period). Charge Class is commonly used to differentiate such corrections from all other charges, including both regularly incurred *charges* and *corrections* to [*open billing periods*](#glossary:open-billing-period).

The ChargeClass column adheres to the following requirements:

* ChargeClass MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* ChargeClass MUST be of type String.
* ChargeClass nullability is defined as follows:

>   * ChargeClass MUST be null when the *charge* does not represent a correction or when it represents a correction within the current *billing period*.

Reviewer: The current formulation of this requirement is **contradictory** to the definition of the ChargeClass column.  The problematic cases are corrections associated with **previous but still open billing periods** (previously referred to as “invoiced billing period”). 

Although this constitutes **a material change**, I suggest updating the normative requirement to **align with the original intent, as outlined in the column definition**.

```suggestion
  * ChargeClass MUST be null when the *charge* does not represent a correction to a previously *closed billing period*.
```

---

***Note**: I hope that in the future we'll revisit the ChargeClass column definition and consider renaming the existing allowed value and adding additional allowed values to differentiate regular charges from corrections in open billing periods and not just corrections to closed billing periods. This would, however, require a thorough discussion.*

# Charge Class

Charge Class indicates whether a [*charge*](#glossary:charge) represents a [*correction*](#glossary:correction) to a previously [*closed billing period*](#glossary:closed-billing-period). Charge Class is commonly used to differentiate such corrections from all other charges, including both regularly incurred *charges* and *corrections* to [*open billing periods*](#glossary:open-billing-period).

## Requirements

ChargeClass adheres to the following requirements:

* ChargeClass MUST be of type String.
* ChargeClass nullability is defined as follows:
  * ChargeClass MUST be null when the *charge* does not represent a correction to a previously *closed billing period*.

>   * ChargeClass MUST NOT be null when the *charge* represents a correction to a previously *closed billing period*.

Reviewer 2: Will the ChargeClass be Null when  the charge pertains to a correction to the current billing period? For e.g. when a one-time invoice related charge ( like subscription purchase) is corrected within the same month

Reviewer: Correct. For open billing periods, many data generators simply replace the data, so we did not introduce a mechanism to track such corrections. (Data generators who want to provide additional information can include it via an `x_` column.)

For corrections to closed billing periods, using ChargeClass = "Correction" and subsequent open billing period, our goal is to ensure better traceability and auditability **at least for closed billing periods**.

Reviewer 2: Can ChargeClass be specified as Correction for all corrections irrespective of closed or open periods.

Reviewer: **Correction** (ChargeClass allowed value) is defined as:

> Correction to a previously closed billing period (e.g., refunds and credit modifications).

From this wording, my understanding is that if a charge does **not** represent a correction to a closed billing period, then `ChargeClass` **MUST** be `null`.

---

We have discussed this concern multiple times (including in this PR)

### Discussion Topic (in this PR)

> The Correction ChargeClass value is currently overloaded. It represents a charge applied to a previously closed billing context, rather than a generic correction. Renaming it to a more descriptive value (e.g. Closed-Billing-Period-Correction, Closed-Billing-Correction, Post-Billing-Correction) would reduce ambiguity.

Attaching [received feedback](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1501#issuecomment-3771356418):

> ## Feedback on "Correction" Terminology (from AI #1830)
> **Decision:** Keep "Correction" as the ChargeClass allowed value (no rename to "Closed Period Correction").
>
> **Rationale:**
> - This is an internal observation, not a recurring practitioner concern
> - Requirements (lines 12–14 of chargeclass.md) are already precise about when to use "Correction"
> - Renaming would be a breaking change to a v1.0 column; appetite for that without a major version is low
> - Enhanced documentation (call-outs, examples, visual aids) adequately addresses semantic clarity
> - Recommendation: Defer sweeping semantic changes to Correction until FOCUS 2.0 (major version change)
>
> **Breaking Change Assessment:** Changing the allowed value would constitute a breaking change

Reviewer 2: Thanks for clarifying that ChargeClass = "Correction" is strictly scoped to closed billing period corrections, and that open-period corrections are handled through data replacement with no correction marker. 

One area worth revisiting in the future: corrections to invoiced charges within an open billing period (e.g., subscription purchases that are invoiced immediately and then corrected). These don't fall under the closed-period definition but are still corrections to charges that have already appeared on an invoice. May be worth considering as the correction handling framework evolves.

Reviewer: Based on the first point, I assume it’s okay to close this thread, so I’m closing it.

For the second point, if the invoice has already been issued, it might not be appropriate to replace the corresponding records. Instead, the correction may need to be delivered using the Append mechanism, Ledger style including reversal charge with ReferenceInvoiceId or possibly updating the status to Void? This feels out of scope for this PR, so I’d **suggest capturing this edge case as a FR** in the backlog to ensure it’s not forgotten.

Also, this reminded me that it might be worth discussing what happens to CostAndUsage charges when InvoiceStatus is set to Void. (I'll ad that as a discussion topic)
